### PR TITLE
[BugFix] Fix merge memo lose best expression

### PR DESCRIPTION
**Pull request by Seaven**
## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

The Bug:
1. Found a best plan, PlanA( group1 -> group2 -> group3 -> group4)
2. Explore group, generate group5
3. found group2 is same as group5, 
    a. we merge group2 & group5
    b. remove groupExpression from group2, because need remove Group2 refs from groupExpression
    c. copy groupExpression to group5, but may group5's statisitcs is different with group2's statistics
5. after merge, need re-enforce group5 expression
6. enforce group5's expression, compute upbound cost by group1, we will get a lower upbound if group1 is join, because join's cost depend on children's statistics(group5's statistics may greate than group2...)
7. then will prune Plan (group5 -> group3 -> group4) because upbound cost
8. then will throw NPE when we extract PlanA( group1 -> group5 -> group3 -> group4)

so...
we can't remove best expression direct when merge group

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 7 / 7 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/Group.java|7|7|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 7 / 7 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/Group.java|7|7|100.00%|[]|

**Comment by liuyehcf** [comment]
「
6. then will prune Plan (group5 -> group3 -> group4) because upbound cost
7. then will throw NPE when we extract PlanA( group1 -> group5 -> group3 -> group4)
」

What is the right plan after this modification?

**Review by liuyehcf — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Group.java [hunk 1]
```diff
@@ -282,6 +279,18 @@ public void mergeGroup(Group other) {
...
+        if (groupExpression.getGroup() == checkGroup) {
+            groupExpression.setGroup(this);
+        }
+
+        for (int i = 0; i < groupExpression.getInputs().size(); i++) {
+            if (groupExpression.inputAt(i) == checkGroup) {
+                groupExpression.getInputs().set(i, this);
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Group.java:292]: I think don't need? with complex enforcer (like window), the lower table will be:
```
1. A enforcer -> B enforcer
2. B enforcer -> group expression
```

I think check child is enough

**Comment by Seaven** [comment]
> 「 6. then will prune Plan (group5 -> group3 -> group4) because upbound cost 7. then will throw NPE when we extract PlanA( group1 -> group5 -> group3 -> group4) 」
> 
> What is the right plan after this modification?

should be `( group1 -> group5 -> group3 -> group4)`

**Review by kangkaisen — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by Seaven** [comment]
@mergify backport branch-2.5

**Comment by Seaven** [comment]
@mergify backport branch-2.4

**Comment by Seaven** [comment]
@mergify backport branch-2.3

**Comment by Seaven** [comment]
@mergify backport branch-2.2

**Pull request merged**